### PR TITLE
Asset redirection via manifest parsing

### PR DESCRIPTION
**Pull request by justinsb**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2769)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Can we rebase? Just merged the cloud provider refactor

**Comment by justinsb** [comment]
Rebased

**Comment by chrislovecnm** [comment]
@justinsb e2e is failing:

```
W0619 22:00:32.984] src/k8s.io/kops/upup/pkg/fi/nodeup/command.go:274: not enough arguments in call to fi.NewDryRunTarget
W0619 22:00:32.984] 	have (io.Writer)
W0619 22:00:32.985] 	want (*assets.AssetBuilder, io.Writer)
W0619 22:00:33.020] make: *** [nodeup-gocode] Error 2
I0619 22:00:33.121] /go/src/k8s.io/kops/Makefile:312: recipe for target 'nodeup-gocode' failed
I0619 22:00:33.523] Makefile:316: recipe for target 'nodeup-dist' failed
```

Also, gofmt is failing as well.

**Comment by justinsb** [comment]
Tests fixed @chrislovecnm

File: pkg/templates/templates.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		t.resources[key] = resource
+	}
+	return nil
+}
+
+func (l *Templates) executeTemplate(key string, d string) (string, error) {
+	t := template.New(key)
+
+	funcMap := make(template.FuncMap)
+	//funcMap["Args"] = func() []string {
```
Comment (chrislovecnm) [inline pkg/templates/templates.go]: Dead code - any reason for it here?

Comment (chrislovecnm) [inline pkg/templates/templates.go]: We can remove later if you want

File: upup/models/cloudup/resources/addons/dns-controller.addons.k8s.io/pre-k8s-1.6.yaml.template [hunk 2]
```diff
@@ -28,7 +28,7 @@ spec:
       hostNetwork: true
       containers:
       - name: dns-controller
-        image: {{ DnsControllerImage }}:1.6.2
+        image: kope/dns-controller:1.6.2
```
Comment (chrislovecnm) [inline upup/models/cloudup/resources/addons/dns-controller.addons.k8s.io/pre-k8s-1.6.yaml.template]: Why was this removed?  This allows for us to override via ENV variable.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 3]
```diff
@@ -171,19 +172,6 @@ func (tf *TemplateFunctions) DnsControllerArgv() ([]string, error) {
 	return argv, nil
 }
 
-// To use user-defined DNS Controller:
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/template_functions.go]: Why we removing this?

Comment (justinsb) [inline upup/pkg/fi/cloudup/template_functions.go]: Moved to AssetBuilder remapImage.  Search for `DNSCONTROLLER_IMAGE` :-)

**Comment by chrislovecnm** [comment]
/assign
/lgtm

You can self merge, once you remove the dead code / code that is commented out
